src/parse.rs

```diff
@@ -43,6 +43,162 @@ impl From<Error> for super::Error {
     }
 }
 
+/// Path to an object in an ObjectStore
+#[non_exhaustive] // permit new variants
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct ObjectStoreUrl {
+    scheme: ObjectStoreScheme,
+    store: Url,
+    path: Path,
+}
+
+impl ObjectStoreUrl {
+    /// Parse a URL into an ObjectStoreUrl
+    ///
+    /// Extracts the relevant ObjectStoreScheme, store URL (for identification),
+    /// and path within the store.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use url::Url;
+    /// # use object_store::{ObjectStoreScheme, ObjectStoreUrl};
+    /// let url: Url = "s3://bucket/path/to/file".parse().unwrap();
+    /// let os_url = ObjectStoreUrl::parse(&url).unwrap();
+    /// assert_eq!(os_url.scheme(), ObjectStoreScheme::AmazonS3);
+    /// assert_eq!(os_url.store().as_str(), "s3://bucket/");
+    /// assert_eq!(os_url.path().as_ref(), "path/to/file");
+    /// ```
+    pub fn parse(url: &Url) -> Result<Self, Error> {
+        let strip_bucket = || Some(url.path().strip_prefix('/')?.split_once('/')?.1);
+        let store_with_bucket = || {
+            let bucket = url
+                .path_segments()
+                .map(|mut segments| segments.next())
+                .flatten()
+                .ok_or(Error::Unrecognised { url: url.clone() })?;
+            Url::parse(&format!(
+                "{}/{}",
+                &url[url::Position::BeforeScheme..url::Position::AfterPort],
+                bucket,
+            ))
+            .map_err(|_| Error::Unrecognised { url: url.clone() })
+        };
+
+        let (scheme, store_url, path) = match (url.scheme(), url.host_str()) {
+            ("file", None) => (
+                ObjectStoreScheme::Local,
+                url[url::Position::BeforeScheme..url::Position::AfterHost].to_string(),
+                &url[url::Position::BeforePath..url::Position::AfterPath],
+            ),
+            ("memory", None) => (
+                ObjectStoreScheme::Memory,
+                url[url::Position::BeforeScheme..url::Position::AfterHost].to_string(),
+                &url[url::Position::BeforePath..url::Position::AfterPath],
+            ),
+            ("s3" | "s3a", Some(_)) => (
+                ObjectStoreScheme::AmazonS3,
+                url[url::Position::BeforeScheme..url::Position::AfterHost].to_string(),
+                url.path(),
+            ),
+            ("gs", Some(_)) => (
+                ObjectStoreScheme::GoogleCloudStorage,
+                url[url::Position::BeforeScheme..url::Position::AfterHost].to_string(),
+                url.path(),
+            ),
+            ("az" | "adl" | "azure" | "abfs" | "abfss", Some(_)) => (
```

Comment:
IIRC there is a pre-existing bug where the bucket is part of the path for some of these, check AzureBuilder